ppx/ppx.ml

```diff
@@ -544,20 +544,46 @@ let makeValue ~loc ~isOptional prop value =
 
 let make_js_props_obj ~loc named_arg_list_with_key_and_ref =
   let label_to_tuple label =
-    let l = Str_label.str label in
-    let id = Exp.ident ~loc {txt= Lident l; loc} in
-    match l with
+    let label_str = Str_label.str label in
+    let id = Exp.ident ~loc {txt= Lident label_str; loc} in
+    match label_str with
     | "key" ->
         [%expr
-          [%e Exp.constant ~loc (Const.string l)]
+          [%e Exp.constant ~loc (Const.string label_str)]
           , inject
               (Js_of_ocaml.Js.Optdef.option
                  (Option.map Js_of_ocaml.Js.string [%e id]) )]
     | "ref" ->
+        [%expr
+          [%e Exp.constant ~loc (Const.string label_str)]
+          , inject (Js_of_ocaml.Js.Optdef.option [%e id])]
+    | _ ->
+        [%expr [%e Exp.constant ~loc (Const.string label_str)], inject [%e id]]
+  in
+  [%expr
+    obj
+      [%e
+        Exp.array ~loc
+          (List.map
+             (fun (label, _, _, _) -> label_to_tuple label)
+             named_arg_list_with_key_and_ref )]]
+
+let make_external_js_props_obj ~loc named_arg_list_with_key_and_ref =
+  let label_to_tuple label =
+    let label_str = Str_label.str label in
+    let id = Exp.ident ~loc {txt= Lident label_str; loc} in
+    match label with
+    | Optional "key" ->
+        [%expr
+          [%e Exp.constant ~loc (Const.string label_str)]
+          , inject
+              (Js_of_ocaml.Js.Optdef.option
+                 (Option.map Js_of_ocaml.Js.string [%e id]) )]
```

Comment:
I am thinking now this is a bit confusing. In most cases at the moment we assume there is no conversion of types in the external definition, but for `key` we will convert from `string` to `js_string`.

I'm not sure what we should do yet, as it's unclear to me if we have to go down the conversion path as you suggest, or leave types untouched. But whatever we do, I think it should be consistent 🤔